src/Framework/ReuseableStringBuilder.cs

```diff
@@ -0,0 +1,346 @@
+#endif
+                }
+                else
+                {
+                    if (returningBuilder.Capacity != returning._borrowedWithCapacity)
+                    {
+                        Debug.Assert(returningBuilder.Capacity > returning._borrowedWithCapacity, "Capacity can only increase");
+
+                        // This builder used more that pre-allocated capacity bracket.
+                        // Let this builder be collected and put new builder, with reflecting bracket capacity, into the pool.
+                        // If we would just return this builder into pool as is, it would allocated new array[capacity] anyway (current implementation of returningBuilder.Clear() does it)
+                        //   and that could lead to unpredictable amount of LOH allocations and eventual LOH fragmentation.
+                        // Bellow implementation have predictable max Log2(MaxBuilderSizeBytes) string builder array re-allocations during whole process lifetime - unless MaxBuilderSizeCapacity is reached frequently.
+                        int newCapacity = SelectBracketedCapacity(returningBuilder.Capacity);
+                        returningBuilder = new StringBuilder(newCapacity);
+                    }
+
+                    returningBuilder.Clear(); // Clear before pooling
+
+                    var oldSharedBuilder = Interlocked.Exchange(ref s_sharedBuilder, returningBuilder);
+                    if (oldSharedBuilder != null)
+                    {
+#if DEBUG
+                        // This can identify in-proper usage from multiple thread or bug in code - Get was reentered before Release.
+                        // User of ReuseableStringBuilder has to make sure that calling method call stacks do not also use ReuseableStringBuilder.
+                        // Look at stack traces of ETW events which contains reported string builder hashes.
+                        MSBuildEventSource.Log.ReusableStringBuilderFactoryUnbalanced(oldHash: oldSharedBuilder.GetHashCode(), newHash: returningBuilder.GetHashCode());
+#endif
+                    }
+#if DEBUG
+                    MSBuildEventSource.Log.ReusableStringBuilderFactoryStop(hash: returningBuilder.GetHashCode(), returningCapacity: returningBuilder.Capacity, returningLength: returningLength, type: returning._borrowedBuilder != returningBuilder ? "return-new" : "return");
+#endif
+                }
+            }
+
+            private static int SelectBracketedCapacity(int requiredCapacity)
+            {
+                foreach (int bracket in s_capacityBrackets)
+                {
+                    if (requiredCapacity <= bracket)
+                    {
+                        return bracket;
+                    }
+                }
+
+                // If user wants bigger capacity than maximum respect it as it could be used as buffer in P/Invoke.
+                return requiredCapacity;
```

Comment:
This can be done without any allocation and with less branching:

```suggestion
                const int minimumCapacity = 0x100; // 256 characters, 512 bytes

                if (requiredCapacity <= minimumCapacity)
                    return minimumCapacity;

                // If user wants bigger capacity than maximum respect it as it could be used as buffer in P/Invoke.
                if (requiredCapacity >= MaxBuilderSizeCapacity)
                    return requiredCapacity;

                // Find next power of two http://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2
                int v = requiredCapacity;

                v--;
                v |= v >> 1;
                v |= v >> 2;
                v |= v >> 4;
                v |= v >> 8;
                v |= v >> 16;
                v++;

                return v;
```

The `s_capacityBrackets` field and the static constructor can then be removed, which IIRC can make working with the type a bit faster as the runtime isn't checking to see if the type has been initialised.

Reply:
Clever!

Reply:
Nice. I liked it. Done